kudu/src/main/java/org/apache/apex/malhar/kudu/scanner/KuduPartitionScannerCallable.java

```diff
@@ -0,0 +1,204 @@
+import org.apache.kudu.client.RowResult;
+import org.apache.kudu.client.RowResultIterator;
+
+import com.datatorrent.lib.util.PojoUtils;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/***
+ * A callable implementation that is responsible for scanning all rows that are scannable from a given
+ * scan token range from a single Kudu tablet. The rows are sent to the input operator buffer so that they
+ * can be emitted as windows progress.
+ */
+public class KuduPartitionScannerCallable<T,C extends InputOperatorControlTuple> implements Callable<Long>
+{
+  private AbstractKuduInputOperator<T,C> operatorUsingThisScanner;
+
+  private KuduPartitionScanAssignmentMeta kuduPartitionScanAssignmentMeta;
+
+  private BlockingQueue<KuduRecordWithMeta<T>> bufferForTransmittingRecords;
+
+  private Class<T> clazzForResultObject;
+
+  private transient KuduClient kuduClientHandle;
+
+  private Map<String,ColumnSchema> tableSchema;
+
+  private Map<String,Object> settersForThisQueryScan;
+
+  private SQLToKuduPredicatesTranslator parsedQuery;
+
+
+  private static final Logger LOG = LoggerFactory.getLogger(KuduPartitionScannerCallable.class);
+
+  public KuduPartitionScannerCallable(AbstractKuduInputOperator<T,C> kuduInputOperator,
+      KuduPartitionScanAssignmentMeta partitionMeta, ApexKuduConnection apexKuduConnection,Map<String,Object> setters,
+      SQLToKuduPredicatesTranslator parsedQueryInstance)
+  {
+    checkNotNull(kuduInputOperator,"Kudu operator instance cannot be null in the kudu scanner thread");
+    checkNotNull(partitionMeta, "Partition metadata cannot be null in kudu scanner thread");
+    checkNotNull(apexKuduConnection,"Kudu connection cannot be null in the kudu scanner thread");
+    checkNotNull(apexKuduConnection,"Setters cannot be null in the kudu scanner thread");
+    checkNotNull(parsedQueryInstance, "parsed Query instance cannot be null");
+    operatorUsingThisScanner = kuduInputOperator;
+    kuduPartitionScanAssignmentMeta = partitionMeta;
+    bufferForTransmittingRecords = kuduInputOperator.getBuffer();
+    clazzForResultObject = kuduInputOperator.getClazzForResultObject();
+    checkNotNull(apexKuduConnection,"Kudu connection cannot be null when initializing scanner");
+    kuduClientHandle = apexKuduConnection.getKuduClient();
+    checkNotNull(kuduClientHandle,"Kudu client cannot be null when initializing scanner");
+    tableSchema = kuduInputOperator.getKuduColNameToSchemaMapping();
+    settersForThisQueryScan = setters;
+    parsedQuery = parsedQueryInstance;
+  }
+
+  public void setValuesInPOJO(RowResult aRow, T payload)
+  {
+    Set<String> columnsUsed = parsedQuery.getKuduSQLParseTreeListener().getListOfColumnsUsed();
+    for (String aColumnName : columnsUsed) {
+      ColumnSchema schemaForThisColumn = tableSchema.get(aColumnName);
+      if (aRow.isNull(aColumnName)) {
+        continue;
+      }
+      switch ( schemaForThisColumn.getType().getDataType().getNumber()) {
```

Comment:
Why not do this with a map of setters that is initialized once?

Reply:
We do not want to throw an error as this "error" state can be reached if there is no valid query that lands on this Apex partition and the client handle socket times out/becomes stale. Hence only logging the stale state and reinitialising the connection using the initially provided connect info object.

Reply:
The reasoning for this is as follows: 

In the SQL expression that is supported by this module, the user can specify a mapping between a Kudu column and the Pojo field name using the "AS" clause.

For example - "Select kuducol1 AS pojofieldone from kudutable where ......"

Since we are letting a query to be supplied for each iteration and the underlying implementation can choose to map the same column in a previous query to a new Pojo Field in the next query.

Hence had to make this parsing logic based on the query change state and not once for all queries.

Reply:
Upon further inspection it appears that this is done for every row, but the result columns remain the same for a given query. I don't think this needs to be changed now. I thought of building a list if extract/set functions that work based on column index instead of names.

Reply:
Perhaps I misunderstood your first comment on this line and hence giving more context. Please let me know if you think there is something more that needs to be done here. 

- The input operator scanner only returns those columns that were asked as part of the SELECT statement ( and does not return all the columns of a table ) and hence is highly dependent on the query being processed. This is for performant reasons as you can guess. 
-  Hence the setters are getting decided only at the beginning of the query processing which you are pointing out.
- This set of setters are actually in another class AbstractKuduInputOperator#extractSettersForResultObject()
- This collection of setter definitions are sent to all threads that are establishing the connection to the Kudu Tablet and retrieving the rows that this Apex operator is responsible for.
- This is done once per query and then passed to the scanner thread as part of its constructor
- In the Scanner thread , after a row is obtained from the Kudu client, only the columns that are returned are used to invoke the setters for those columns.
- We are not defining the setters but only calling the setter to set the value from the Kudu Row into the Pojo Field.
- A POJO field can be a superset of the Kudu columns in terms of mapping. I was thinking that will result in better usability of POJOs as we do not need a separate POJO across operators if required.

Reply:
Yes, the columns that need to be processed depend on the query and not what columns a given table has. This is similar to the JDBC case where you use the result set metadata and not the table metadata. My idea was that you could build a list of functions that handle both, get from result and set on object that are subsequently just referenced by index for processing the rows. Not in the scope of this PR though, we can discuss it separately.